src/content/en/updates/posts/2016/07/infinite-scroller.md

```diff
@@ -0,0 +1,246 @@
+
+The technical challenge behind a infinite scroller, however, is harder than it seems. The range of problems you encounter when you want to do The Right Thing™ is vast. It starts with simple things like the links in the footer becoming practically unreachable because content keeps pushing the footer away. But the problems get harder. How do you handle a resize event when someone turns their phone from portrait to landscape or how do you prevent your phone from grinding to a painful halt when the list gets too long?
+
+#The Right Thing™
+We thought that was reason enough to come up with a reference implementation that shows a way to tackle all these problems in a reusable way while maintaining performance standards.
+
+We are going to utilize 3 techniques to achieve our goal: DOM recycling, tombstones and scroll anchoring.
+
+Our demo case is going to be a Hangouts-like chat window where we can scroll through the messages. The first thing we need is an infinite source of chat messages. Technically, none of the infinite scrollers out there are *truly*   infinite, but with the amount of data that is available to get pumped into these scrollers they might as well be. For simplicity’s sake we will just hard-code a set of chat messages and pick message, author and occasional image attachment at random with a sprinkle of artificial delay to behave a little bit more like the real network.
+
+<img src="/web/updates/images/2016/07/infinite-scroller/screenshot.png">
+
+#DOM recycling
+DOM recycling is a (heavily underutilized) technique to keep the DOM node count low. Admittedly, DOM nodes themselves are cheap, but they are not free, as each of them adds extra cost in memory, layout, style and paint. Low-end devices will get noticeably slower if not completely unusable if the website has too big of a DOM to manage. Also keep in mind that every relayout and reapplication of your styles – a process that is triggered whenever a class is added or removed from a node – grows more expensive with a bigger DOM. Recycling your DOM nodes means that we are going to keep the total number of DOM nodes considerably lower, making all these processes faster.
+
+The first hurdle is the scrolling itself. Since we will only have a tiny subset of all available items in the DOM at any given time, we need to find another way to make the browser’s scrollbar properly reflects the amount content that is theoretically there. We will use a 1px by 1px sentinel element with a transform to force the container element – the runway – to have the desired height. We have to be careful to keep the layer of the runway completely empty, though. No background color, nothing. The second the layer contains more than that the browser’s optimizations won’t skip processing that layer and we will have to store a texture on our graphics card that has a height of a couple of hundred thousand pixels. Definitely not viable on a mobile device.
```

Comment:
Runway - You should define your terms with a reference to your diagram before throwing them at your reader.